sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -339,11 +339,34 @@ case class DataSource(
         dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
       case (_: SchemaRelationProvider, None) =>
         throw new AnalysisException(s"A schema needs to be specified when using $className.")
-      case (dataSource: RelationProvider, Some(schema)) =>
+      case (dataSource: RelationProvider, Some(specifiedSchema)) =>
         val baseRelation =
           dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
-        if (baseRelation.schema != schema) {
-          throw new AnalysisException(s"$className does not allow user-specified schemas.")
+        val persistentSchema = baseRelation.schema
+        val persistentSize = persistentSchema.size
+        val specifiedSize = specifiedSchema.size
+        if (persistentSize == specifiedSize) {
+          val (persistentFields, specifiedFields) = persistentSchema.zip(specifiedSchema)
```

Comment:
If we're going to improve such error message case across the codebase, we might also think about having a common method (maybe something called `assertEquality` in `StructType`?) that checks each type recursively and shows a better message. Can we at least have a private method for this case in the future?

Reply:
Not sure whether we'd require this similar functionality in some cases in the future. But,  maybe, we could still give it a try.

Reply:
Yeah, I think it wont handle nested cases. There are other data sources that support nested schema and the current code tells only root columns.

Reply:
Also, there are many cases to show better error messages like this. E.g., StructType.merge or  `_merge_type` in Python's schema inference (https://github.com/apache/spark/blob/master/python/pyspark/sql/types.py#L1097-L1111)

Reply:
see https://github.com/apache/spark/pull/19792 or https://github.com/apache/spark/pull/18521 as an example.

Reply:
Hi @HyukjinKwon , after discussing with wenchen offline, we decide not to make it too complicated here. If schemas are detected not match, we simply show the whole schema to user rather than those mismatched fields as previously did. Please see de036b6.